addons/hi-res-thumbnails/userscript.js

```diff
@@ -0,0 +1,77 @@
+export default async function ({ addon, console }) {
+  const DEBUG = true;
+
+  main: while (true) {
+    const image = await addon.tab.waitForElement("img", {
+      markAsSeen: true,
+    });
+
+    const lazy = image.classList.contains("lazy");
+    const src = lazy ? image.dataset.original : image.src;
+
+    let width, height, newSrc;
+
+    const cdn2 = src.match(/(.*\/get_image\/.*?\/[0-9]+?_)([0-9]+?)x([0-9]+?)(\.[a-z]+)/);
+
+    if (cdn2) {
+      width = cdn2[2];
+      height = cdn2[3];
+    } else {
+      continue main;
+    }
+
+    if (DEBUG) {
+      image.dataset.debugOriginalWidth = width;
+      image.dataset.debugOriginalHeight = height;
+    }
+
+    if (addon.settings.get("fixLowRes"))
+      fixLowRes: {
+        // Note this returns a value in CSS pixel units, i.e. it doesn't take into
+        // account browser scaling or high-DPI screens. That's good, because we
+        // want to leave the actual resolution-increase factor completely up to
+        // the addon user/settings!
+        const rect = image.getBoundingClientRect();
+
+        // Use greater of both scales to choose the dimension which needs the most
+        // scaling and to maintain the ratio between the rendered and loaded
+        // dimensions.
+        const scale = Math.max(rect.width / width, rect.height / height);
+
+        if (scale > 1) {
+          width *= scale;
+          height *= scale;
+        }
```

Comment:
One point of discussion re: implementation of this resolution fix: maybe it makes sense to *down*scale images which are by default loaded in at a higher resolution than they're actually displayed at? (Before applying the user-specified multiplier, of course.)

The main example I found of this is on the new studio pages, where for some reason thumbnails are loaded at 480x360, even though they're displayed at only about 217x203 in CSS pixels on a 1920p screen. This addon would actually save a little data usage if you set the res setting to 100% and had this fix enabled, then, because it would fetch the images at 217x203 instead of 480x360!

Although apparently, studio thumbnails load a higher res because the rendered size is semi-flexible, i.e. at thinner displays (e.g. by shrinking the window) it'll only show two projects per row, and this can get the thumbnails to around 345x260 CSS pixels. Still less than 480x360, though not by as much. Just the same, it points out a flaw with downscaling to the rendered size: if the image later dynamically *expands*, it'll look lower resolution than the original developers were giving excess dimensions as a way of intentionally allowing for!